src/core/lib/event_engine/cf_engine/dns_service_resolver.cc

```diff
@@ -0,0 +1,222 @@
+  }
+
+  requests_.try_emplace(
+      sdRef, DNSServiceRequest{
+                 std::move(on_resolve), static_cast<uint16_t>(port), {}});
+}
+
+/* static */
+void DNSServiceResolverImpl::ResolveCallback(
+    DNSServiceRef sdRef, DNSServiceFlags flags, uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode, const char* hostname,
+    const struct sockaddr* address, uint32_t ttl, void* context) {
+  GRPC_EVENT_ENGINE_DNS_TRACE(
+      "DNSServiceResolverImpl::ResolveCallback: sdRef: %p, flags: %x, "
+      "interface: %d, errorCode: %d, hostname: %s, addressFamily: %d, ttl: "
+      "%d, "
+      "this: %p",
+      sdRef, flags, interfaceIndex, errorCode, hostname, address->sa_family,
+      ttl, context);
+
+  // no need to increase refcount here, since ResolveCallback and Shutdown is
+  // called from the serial queue and it is guarenteed that it won't be called
+  // after the sdRef is deallocated
+  auto that = static_cast<DNSServiceResolverImpl*>(context);
+
+  grpc_core::ReleasableMutexLock lock(&that->request_mu_);
+  auto request_it = that->requests_.find(sdRef);
+  GPR_ASSERT(request_it != that->requests_.end());
+  auto& request = request_it->second;
+
+  if (errorCode != kDNSServiceErr_NoError &&
+      errorCode != kDNSServiceErr_NoSuchRecord) {
+    request.on_resolve(absl::UnknownError(absl::StrFormat(
+        "address lookup failed for %s: errorCode: %d", hostname, errorCode)));
+    that->requests_.erase(request_it);
+    DNSServiceRefDeallocate(sdRef);
+    return;
+  }
+
+  // set received ipv4 or ipv6 response, even for kDNSServiceErr_NoSuchRecord
+  if (address->sa_family == AF_INET) {
+    request.has_ipv4_response = true;
+  } else if (address->sa_family == AF_INET6) {
+    request.has_ipv6_response = true;
+  }
+
+  // collect results if there is no error (not kDNSServiceErr_NoSuchRecord)
+  if (errorCode == kDNSServiceErr_NoError) {
+    request.result.emplace_back(address, address->sa_len);
+    auto& resolved_address = request.result.back();
+    if (address->sa_family == AF_INET) {
+      (const_cast<sockaddr_in*>(
+           reinterpret_cast<const sockaddr_in*>(resolved_address.address())))
+          ->sin_port = htons(request.port);
+    } else if (address->sa_family == AF_INET6) {
+      (const_cast<sockaddr_in6*>(
+           reinterpret_cast<const sockaddr_in6*>(resolved_address.address())))
+          ->sin6_port = htons(request.port);
+    }
+
+    GRPC_EVENT_ENGINE_DNS_TRACE(
+        "DNSServiceResolverImpl::ResolveCallback: "
+        "sdRef: %p, hostname: %s, addressPort: %s, this: %p",
+        sdRef, hostname,
+        ResolvedAddressToString(resolved_address).value_or("ERROR").c_str(),
+        context);
+  }
+
+  if (!(flags & kDNSServiceFlagsMoreComing) && request.has_ipv4_response &&
+      request.has_ipv6_response) {
+    if (request.result.empty()) {
+      request.on_resolve(absl::NotFoundError(absl::StrFormat(
+          "address lookup failed for %s: Domain name not found", hostname)));
```

Comment:
Would it be cleaner to condition on `kDNSServiceErr_NoSuchRecord rather than request.result.empty()?

Reply:
This callback will be called at least two times which can be:
1. callback with an IPv4 address
2. callback with kDNSServiceErr_NoSuchRecord to indicate no IPv6 address

So in the second call of callback, although I got no records error, still need to return the IPv4 address.
This can also happen in reversed order where NoSuchRecord comes in the first call of the callback, I need to keep waiting for the result to come from the other stack.

Reply:
oh ok! thanks for explaining! (btw if you think this logic is not obvious to understand, please consider putting some comments for future understandings :))

Reply:
Yeah, it also took me a while to understand how it works as there isn't much document about it.
Let me see how I can better describe it for the resolving flow, which maybe in a follow up PR

Reply:
It's ok. No big deal.. Just do it only if you see fit.. Thanks for explaining! :)